src/System.Linq.Expressions/tests/IndexExpression/IndexExpressionTests.cs

```diff
@@ -64,6 +66,580 @@ public static void ToStringTest()
+        {
+            PropertyInfo prop = typeof(Dictionary<int, int>).GetProperty("Item");
+            DefaultExpression dict = Expression.Default(typeof(Dictionary<int, int>));
+            ConstantExpression index = Expression.Constant(0L);
+            Assert.Throws<ArgumentException>("arguments[0]", () => Expression.Property(dict, prop, index));
+        }
+
+        [Theory, ClassData(typeof(CompilationTypes))]
+        public void CallWithLambdaIndex(bool useInterpreter)
+        {
+            // An exception to the rule against unassignable indices, lamdba expressions
+            // can be automatically quoted.
+            PropertyInfo prop = typeof(Dictionary<Expression<Func<int>>, int>).GetProperty("Item");
+            Expression<Func<int>> index = () => 2;
+            ConstantExpression dict = Expression.Constant(new Dictionary<Expression<Func<int>>, int>{{index, 9}});
+            Func<int> f = Expression.Lambda<Func<int>>(Expression.Property(dict, prop, index)).Compile(useInterpreter);
+            Assert.Equal(9, f());
+        }
+
+        [Theory, ClassData(typeof(CompilationTypes))]
+        public void CallWithIntAndLambdaIndex(bool useInterpreter)
+        {
+            PropertyInfo prop = typeof(IntAndExpressionIndexed).GetProperty("Item");
+            ConstantExpression instance = Expression.Constant(new IntAndExpressionIndexed());
+            Expression<Action> index = Expression.Lambda<Action>(Expression.Empty());
+            ConstantExpression intIdx = Expression.Constant(0);
+            Func<bool> f = Expression.Lambda<Func<bool>>(Expression.Property(instance, prop, intIdx, index)).Compile(useInterpreter);
+            Assert.True(f());
+        }
+
+        [Fact]
+        public void TryIndexedAccessNonIndexedProperty()
+        {
+            ConstantExpression instance = Expression.Constant("");
+            PropertyInfo prop = typeof(string).GetProperty(nameof(string.Length));
+            ConstantExpression index = Expression.Constant(0);
+            Assert.Throws<ArgumentException>("indexer", () => Expression.Property(instance, prop, index));
+        }
+
+        [Theory, ClassData(typeof(CompilationTypes))]
+        public void OverloadedIndexer(bool useInterpreter)
+        {
+            ConstantExpression instance = Expression.Constant(new OverloadedIndexers());
+            ConstantExpression index = Expression.Constant("");
+            Expression<Func<int>> exp = Expression.Lambda<Func<int>>(Expression.Property(instance, "Item", index));
+            Func<int> f = exp.Compile(useInterpreter);
+            Assert.Equal(2, f());
+        }
+
+        [Fact]
+        public void OverloadedIndexerBothMatch()
+        {
+            ConstantExpression instance = Expression.Constant(new OverloadedIndexersBothMatchString());
+            ConstantExpression index = Expression.Constant("");
+            Assert.Throws<InvalidOperationException>(() => Expression.Property(instance, "Item", index));
```

Comment:
I think this should really be ArgumentException, but isn't worth the potential of breaking to change that.